TypeError on poly.ground_roots() with Flint

This code used to work with sympy 1.13.3. With sympy 1.14.0 and python-flint 0.7.1 it no longer does. Also even 1.14.0rc1 fails.

```python
    from sympy import symbols, FF, Poly

    ax = symbols("α")
    field = FF(340282366762482138434845932244680310783)
    rhs = Poly(
        ax**3
        + field(55811479606114091134440010920619299102) * ax
        + field(126188322377389722996253562430093625949),
        ax,
        domain=field,
    )
    rhs.ground_roots()
    # raises TypeError:
    # flint/types/fmpz_mod_poly.pyx:321: TypeError
```
Here is the traceback.
```python-traceback
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../virt/lib/python3.13/site-packages/sympy/polys/polytools.py:3772: in ground_roots
    for factor, k in f.factor_list()[1]:
../virt/lib/python3.13/site-packages/sympy/polys/polytools.py:3377: in factor_list
    coeff, factors = f.rep.factor_list()
../virt/lib/python3.13/site-packages/sympy/polys/polyclasses.py:2316: in factor_list
    factors = f._sort_factors(factors)
../virt/lib/python3.13/site-packages/sympy/polys/polyclasses.py:2337: in _sort_factors
    return [ (to_dup_flint(g), k) for g, k in factors ]
../virt/lib/python3.13/site-packages/sympy/polys/polyclasses.py:2336: in <lambda>
    to_dup_flint = lambda g: f.from_rep(f._cls(g[::-1]), f.dom)
../virt/lib/python3.13/site-packages/sympy/polys/polyclasses.py:1786: in <lambda>
    _cls = lambda e: __cls(e, c)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   TypeError

flint/types/fmpz_mod_poly.pyx:321: TypeError
```

--- a/.mailmap
+++ b/.mailmap
@@ -768,6 +768,7 @@ James Titus <titusjames299@gmail.com> <jamestitus299@gmail.com>
 James Titus <titusjames299@gmail.com> <titusjames299@gmail.com>
 James Whitehead <whiteheadj@gmail.com> jcwhitehead <whiteheadj@gmail.com>
 Jan Jancar <johny@neuromancer.sk> J08nY <johny@neuromancer.sk>
+Jan Jancar <johny@neuromancer.sk> Ján Jančár <J08nY@users.noreply.github.com>
 Jan Kruse <janckruse@t-online.de>
 Jan-Philipp Hoffmann <sonntagsgesicht@icloud.com> Jan-Philipp Hoffmann <90828785+jan-philipp-hoffmann@users.noreply.github.com>
 Jared Lumpe <mjlumpe@gmail.com> Michael Jared Lumpe <mjlumpe@gmail.com>

--- a/sympy/polys/polyclasses.py
+++ b/sympy/polys/polyclasses.py
@@ -1764,7 +1764,7 @@ class DUP_Flint(DMP):
             return flint.fmpz_poly
         elif dom.is_QQ:
             return flint.fmpq_poly
-        elif dom.is_FF:
+        elif dom.is_FF and dom._is_flint:
             return dom._poly_ctx
         else:
             raise RuntimeError("Domain %s is not supported with flint" % dom)
@@ -1773,19 +1773,7 @@ class DUP_Flint(DMP):
     def from_rep(cls, rep, dom):
         """Create a DMP from the given representation. """
 
-        if dom.is_ZZ:
-            assert isinstance(rep, flint.fmpz_poly)
-            _cls = flint.fmpz_poly
-        elif dom.is_QQ:
-            assert isinstance(rep, flint.fmpq_poly)
-            _cls = flint.fmpq_poly
-        elif dom.is_FF:
-            assert isinstance(rep, (flint.nmod_poly, flint.fmpz_mod_poly))
-            c = dom.characteristic()
-            __cls = type(rep)
-            _cls = lambda e: __cls(e, c)
-        else:
-            raise RuntimeError("Domain %s is not supported with flint" % dom)
+        _cls = cls._get_flint_poly_cls(dom)
 
         obj = object.__new__(cls)
         obj.dom = dom

--- a/sympy/polys/tests/test_polytools.py
+++ b/sympy/polys/tests/test_polytools.py
@@ -4068,3 +4068,17 @@ def test_issue_20985():
     w, R = symbols('w R')
     poly = Poly(1.0 + I*w/R, w, 1/R)
     assert poly.degree() == S(1)
+
+
+def test_issue_28156():
+    from sympy.core.symbol import symbols
+
+    ax = symbols("a")
+    field = FF(340282366762482138434845932244680310783)
+    rhs = Poly(
+        ax**3 + field(3) * ax + field(5),
+        ax,
+        domain=field,
+    )
+    roots = rhs.ground_roots()
+    assert roots

